### ISSUE TITLE
How to change surface type now?

### ISSUE THREAD

**Issue by EvgenyK-SO** [ladybug-tools/spider-gbxml-tools#46]
Hello.
With the current Maevia version I don't see the ability to change the surface type on the right panel, i.e. from Air to Exterior Wall.
How can I do that?
Thank you.

**Comment by theo-armour** [comment]
@EK-CEL 

Sorry for slow reply to this issue. Somehow it does not seem to have made it to my mailbox. Or more likely I deleted the notice by mistake.

There are several answers to your question. It depends on the type of error

In Maevia v0.17.O2 in the Fixer menu > Fix air surfaces with single adjacent space. This is a basic check and needs more testing.

In Maevia r15.3.5 there are the more sophisticated  InteriorWall and InteriorFloor on exterior items. This code with 3D ray casting should reappear in upcoming releases and handle air surfaces as well.

The overarching strategy is to leave behind humans searching for and fixing dozens of errors one error at a time to letting the code fix all the issues with a single click.

For this we will need test cases and descriptions of the logic to follow.

**Comment by EvgenyK-SO** [comment]
@theo-armour 
I am wondering - for a building like this:
![image](https://user-images.githubusercontent.com/39136340/62947968-82a19780-bdb1-11e9-9c33-66eb4acf67f9.png)

How can the Spider automatically recognize what is the void in the building: shaft or courtyard?

**Comment by theo-armour** [comment]
@EK-CEL 

See R15

https://www.ladybug.tools/spider-gbxml-tools/spider-gbxml-viewer/r15/spider-gbxml-viewer15.html

Uses ray-casting. Any ray emanating from a wall that that passes through an even number of external surfaces must itself be external.

Will be in v0.17 - and simplified - before too long

**Comment by EvgenyK-SO** [comment]
@theo-armour 
Sorry, I probably didn't describe problem clearly.
Let's assume that the hole in the building is a shaft. Quite often our designers don't put any space for a shaft, it is just an empty sport in the model. Unfortunately Revit not always can properly determine the surface type properly. It may happen that two of the surfaces appear as internal and other two as external in the exported gbXML file. We need to fix that model. We know (because of the shaft) that the "external" surfaces are incorrect and they should be converted to internal type. But if we assume that the hole is a courtyard, then the internal surfaces are incorrect, and they should be changed to internal. So the target result for the hole boundary surface type (internal/external) depends of what it represents - shaft or courtyard. And that knowledge exists in our mind only. Spider cannot guess that having just a gbXML model.

**Comment by michaldengusiak** [comment]
@EK-CEL I got your point and this is correct... this is reason  why we edit models and add missing and ask always Architect to put rooms for all spaces. This automation will be very complicated as to get it 100% correct gbXML we will require to add missing space/zone and close from top and bottom.  How do we recognize what is shaft... I guess the most complicated here is geometrical modification but .... I wish we can fix all architect errors with one click ... :)

**Comment by EvgenyK-SO** [comment]
Unfortunately for a real project we always have "holes".
![image](https://user-images.githubusercontent.com/39136340/63028712-5bb09780-be7d-11e9-967e-0372ba0bf02c.png)

We can place spaces for the "green" shafts, they are relatively large. And again, we cannot guarantee that input. But there are tons of "red" tiny holes with area up to 1 square meter, or a piping chase, which nobody is going to care about. We should not expect to have spaces there. And such cases should be fixed automatically. 
Probably we should establish hole size thresholds: area less then 10 sq.m and width less then 600 mm. All the surfaces surrounding such hole should be converted to internal walls. If the hole size is bigger then it is a courtyard and the walls should be healed accordingly. It is just a hypotheses. :)

**Comment by EvgenyK-SO** [comment]
@theo-armour @mdengusiak  
I don't see any activities for the project. It bothers me. Our designers still use Version 14. They need to change surface type and with the that version they pick surfaces one by one. It is so annoying. And it is so desirable that the Spider can help.
Frankly, I don't believe in a fully automatic fix. I think the tool can fix most of issues, but still a manual fix is required after that. And sometime automation can bring more harm than help. I believe a manual fix tool should be a part of the Spider in any development scenario.
Gentlemen, I am begging you - please make multi-selection and multi-parameter changing. I believe it can be coded quite quickly. And it will be tremendously helpful. Please. I really hope the project is still alive.

**Comment by michaldengusiak** [comment]
I agree in fact I am still using R12, as a manual workaround is still needed on most projects... let's see what we can do with @theo-armour in new version

**Comment by theo-armour** [comment]
@EK-CEL @michaldengusiak @mostaphaRoudsari @chriswmackey @AntoineDao 

Evgeny and Michal: thank you for raising these issues.

> Our designers still use Version 14. 
> I agree in fact I am still using R12

I am glad you are happy with my 2017 and 2018 efforts. 

Please may I go away now? ;-)


> I don't see any activities for the project. It bothers me. Our designers still use Version 14

Did you ever consider improving your GitHub skill levels?

Since September 2019 I have been reporting Spider gbXML efforts here:

https://github.com/GreenBuildingXML/GreenBuildingXML.github.io/issues

See issues "closed' to verify.

If you need a more detailed look at the Spider gbXML Viewer, here is a good link:

https://github.com/ladybug-tools/spider-gbxml-tools/tree/master/spider-gbxml-viewer:

If you would like to check out if I am being lazy, go here and look for the white boxes:

https://github.com/theo-armour

> They need to change surface type and with the that version they pick surfaces one by one. **It is so annoying.**

Writing annoying software is the best. It means that there are users who are smart enough to understand they are being tortured. ;-)

The interesting question is will the next revision be even more annoying or will the users stage a revolution? ;-)

My personal feeling about 2019 was that I am 72 and my time on Spider projects was a waste of one of the few remaining years of my productive life.

For example, here are ten Spider projects of interest that I worked on - not including the gbXML Viewer:

* https://www.ladybug.tools/spider-gbxml-tools/sandbox/spider-gbxml-mapper/
* https://www.ladybug.tools/spider-gbxml-tools/sandbox/spider-gbxml-collaborator
* https://www.ladybug.tools/spider-gbxml-tools/sandbox/spider-gbxml-manifold
* https://www.ladybug.tools/spider-gbxml-tools/sandbox/fetch-assemble-save
* https://www.ladybug.tools/spider-gbxml-fixer
* https://ladybug-tools.github.io/spider/cookbook/epw-data-onebuilding/
* https://www.ladybug.tools/spider/cookbook/epw-data-onebuilding/
* https://www.ladybug.tools/spider/cookbook/gbxml-diff-engine/v-0-01-00/gbxml-diff-engine.html
* https://www.ladybug.tools/spider/cookbook/manifold/manifold-04/
* https://www.ladybug.tools/spider/cookbook/square-to-hemisphere/v-0-02-00/square-to-hemisphere.html

For a variety of good reasons none of these projects - and a number of others - went anywhere.

In almost all the cases the reasons for lack of movement were the same: somebody said they wanted something but no test cases appeared, follow-up was almost non-existent and no spec was ever developed. There was no actual teamwork.

My efforts on the Spider gbXML Viewer project were similar. Often I was sent a broken file and asked how to fix it because there was a due date approaching. Great architecture does not arise merely from being good about putting out fires.

It would be nice if the 2020 objectives for the Spider project will be about enabling buildings are are more efficient, cost effective and pleasurable. And, fingers crossed, we build things that are less annoying.

**Comment by EvgenyK-SO** [comment]
@theo-armour 
I greatly appreciate your contribution into the Spider project. Looking at the list of the projects you participated in it would be wrong even to mention your laziness. ;-).
But I can say about my laziness ;-). I have a full time job and when I commute home I don't have any energy for learning the GitHub. I understand nearly nothing in programming and I am not sure that I can contribute any quality code into the project. Maybe I should try ;-).

Thank you for the links, they are new for me. I will take a look what happens on other forks.

I wish you to find a fulfilling objective for your activity.

**Comment by theo-armour** [comment]
@EK-CEL 

> I greatly appreciate your contribution into the Spider project

Thank you for your kind words. I know that both you and Michal have been my best supporters.

 > I don't have any energy for learning the GitHub

There is no requirement for you to learn GitHub. But what may be of interest is being able to spot activity and interesting things happening around Ladybug Tools. And that's what the links I sent are for.

> I wish you to find a fulfilling objective for your activity.

Thank you. I have started drafting Spider objectives for 2020. First pass was here (at bottom):

https://github.com/ladybug-tools/community/issues/5

I should have a second draft up soon.
